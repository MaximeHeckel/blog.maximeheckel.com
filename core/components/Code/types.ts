import { Language } from 'prism-react-renderer';
import React from 'react';

export type PrePropsType = {
  // children: {
  //   props: {
  //     metastring: string;
  //     mdxType?: string;
  //     className?: string;
  //     children: string;
  //   };
  // };
  children?: React.ReactNode;
};

export interface CodeBlockProps {
  codeString: string;
  language: Language;
  metastring: string | null;
}

export interface HighlightedCodeTextProps {
  codeString: string;
  language: Language | 'glsl';
  highlightLine?: (index: number) => boolean;
}
