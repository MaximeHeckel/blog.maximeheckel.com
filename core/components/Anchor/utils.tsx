import { ArrowPosition } from './types';

/**
 * Base 64 equivalent of the ArrowIcon, TwitterIcon and GithubIcon with width and height set to "var(--size, 1.05em)"
 */
const Arrow =
  'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSJ2YXIoLS1zaXplLCAxLjA1ZW0pIiBoZWlnaHQ9InZhcigtLXNpemUsIDEuMDVlbSkiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgc3Ryb2tlPSJjdXJyZW50Q29sb3IiIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIj48bGluZSB4MT0iNSIgeTE9IjEyIiB4Mj0iMTkiIHkyPSIxMiI+PC9saW5lPjxwb2x5bGluZSBwb2ludHM9IjEyIDUgMTkgMTIgMTIgMTkiPjwvcG9seWxpbmU+PC9zdmc+';

const Twitter =
  'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSJ2YXIoLS1zaXplLCAxLjA1ZW0pIiBoZWlnaHQ9InZhcigtLXNpemUsIDEuMDVlbSkiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgc3Ryb2tlPSJjdXJyZW50Q29sb3IiIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIiBjbGFzcz0iZmVhdGhlciBmZWF0aGVyLXR3aXR0ZXIiPjxwYXRoIGQ9Ik0yMyAzYTEwLjkgMTAuOSAwIDAgMS0zLjE0IDEuNTMgNC40OCA0LjQ4IDAgMCAwLTcuODYgM3YxQTEwLjY2IDEwLjY2IDAgMCAxIDMgNHMtNCA5IDUgMTNhMTEuNjQgMTEuNjQgMCAwIDEtNyAyYzkgNSAyMCAwIDIwLTExLjVhNC41IDQuNSAwIDAgMC0uMDgtLjgzQTcuNzIgNy43MiAwIDAgMCAyMyAzeiI+PC9wYXRoPjwvc3ZnPg==';

const Github =
  'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSJ2YXIoLS1zaXplLCAxLjA1ZW0pIiBoZWlnaHQ9InZhcigtLXNpemUsIDEuMDVlbSkiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgc3Ryb2tlPSJjdXJyZW50Q29sb3IiIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIiBjbGFzcz0iZmVhdGhlciBmZWF0aGVyLWdpdGh1YiI+PHBhdGggZD0iTTkgMTljLTUgMS41LTUtMi41LTctM20xNCA2di0zLjg3YTMuMzcgMy4zNyAwIDAgMC0uOTQtMi42MWMzLjE0LS4zNSA2LjQ0LTEuNTQgNi40NC03QTUuNDQgNS40NCAwIDAgMCAyMCA0Ljc3IDUuMDcgNS4wNyAwIDAgMCAxOS45MSAxUzE4LjczLjY1IDE2IDIuNDhhMTMuMzggMTMuMzggMCAwIDAtNyAwQzYuMjcuNjUgNS4wOSAxIDUuMDkgMUE1LjA3IDUuMDcgMCAwIDAgNSA0Ljc3YTUuNDQgNS40NCAwIDAgMC0xLjUgMy43OGMwIDUuNDIgMy4zIDYuNjEgNi40NCA3QTMuMzcgMy4zNyAwIDAgMCA5IDE4LjEzVjIyIj48L3BhdGg+PC9zdmc+';

export const getIconString = (
  href?: string,
  arrow?: ArrowPosition
): string | null => {
  switch (true) {
    case typeof arrow !== 'undefined':
      return Arrow;
    case href && href.includes('twitter.com'):
      return Twitter;
    case href && href.includes('github.com'):
      return Github;
    default:
      // Improve: Add default link icon here
      return null;
  }
};
