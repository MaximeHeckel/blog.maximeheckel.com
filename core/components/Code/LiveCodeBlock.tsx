import {
  motion,
  useAnimation,
  useMotionValue,
  useTransform,
  AnimatePresence,
  LayoutGroup,
} from 'framer-motion';
import { styled, Pill } from '@maximeheckel/design-system';
import { PrismTheme } from 'prism-react-renderer';
import React from 'react';
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live';
import { CodeBlockProps } from './types';

const LiveCodeBlock = (props: CodeBlockProps) => {
  const { codeString, live, render } = props;

  const scope = {
    motion,
    AnimatePresence,
    LayoutGroup,
    useAnimation,
    useMotionValue,
    useTransform,
    styled,
    Pill,
    React,
  };

  const customTheme = {
    styles: [],
    plain: {},
  } as PrismTheme;

  if (live) {
    return (
      <LiveProvider
        theme={customTheme}
        code={codeString}
        scope={scope}
        noInline={true}
      >
        <StyledLiveCodeWrapper>
          <StyledPreviewWrapper withEditor>
            <LivePreview />
            <StyledErrorWrapper>
              <LiveError />
            </StyledErrorWrapper>
          </StyledPreviewWrapper>
          <StyledEditorWrapper>
            <LiveEditor />
          </StyledEditorWrapper>
        </StyledLiveCodeWrapper>
      </LiveProvider>
    );
  }

  if (render) {
    return (
      <LiveProvider code={codeString} scope={scope} noInline={true}>
        <StyledLiveCodeWrapper>
          <StyledPreviewWrapper>
            <LivePreview />
          </StyledPreviewWrapper>
        </StyledLiveCodeWrapper>
      </LiveProvider>
    );
  }

  return null;
};

const StyledLiveCodeWrapper = styled('div', {
  position: 'relative',
  backdropFilter: 'blur(6px)',
  display: 'flex',
  alignItems: 'center',
  marginBottom: '32px',
  borderRadius: 'var(--border-radius-2)',
  overflow: 'hidden',

  '@media (max-width: 750px)': {
    display: 'block',
  },

  '@media (max-width: 1200px)': {
    /**
     * Make it fullbleed!
     */
    width: '100vw',
    left: '50%',
    right: '50%',
    marginLeft: '-50vw',
    marginRight: '-50vw',
    borderRadius: '0px',
  },

  '@media (min-width: 1200px)': {
    position: 'relative',
    width: 'calc(100% + 200px)',
    marginLeft: '-100px',
    marginRight: '-100px',
  },
});

const StyledEditorWrapper = styled('div', {
  flex: '50 1 0%',
  height: '100%',
  maxHeight: '600px',
  overflow: 'auto',
  margin: '0',

  'textarea:focus': {
    outline: 'none',
  },

  'pre, textarea': {
    backgroundColor: 'var(--code-snippet-background) !important',
    fontFamily: 'var(--font-mono) !important',
    fontSize: 'var(--font-size-1) !important',
    lineHeight: '26px !important',
  },
});

const StyledPreviewWrapper = styled('div', {
  maxHeight: '600px',
  flex: '50 1 0%',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  backgroundColor: 'var(--maximeheckel-colors-emphasis)',
  overflow: 'hidden',

  '@media (max-width: 750px)': {
    borderRadius: '0px !important',
  },

  variants: {
    withEditor: {
      true: {
        minHeight: '600px',
      },
      false: {
        minHeight: '300px',
      },
    },
  },
});

const StyledErrorWrapper = styled('div', {
  color: 'var(--maximeheckel-colors-typeface-secondary)',
  maxWidth: '300px',

  pre: {
    padding: '15px',
    marginBottom: '0px',
  },
});

export default LiveCodeBlock;
