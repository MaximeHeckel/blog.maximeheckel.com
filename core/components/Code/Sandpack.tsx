import {
  SandpackProvider,
  SandpackPreview,
  SandpackCodeEditor,
  SandpackLayout,
  SandpackPredefinedTemplate,
} from '@codesandbox/sandpack-react';
import { Shadows, styled } from '@maximeheckel/design-system';
import setupFiles from './SandpackSetupFiles';

// Default Theme
const theme = {
  colors: {
    hover: 'var(--maximeheckel-colors-brand)',
    clickable: 'var(--maximeheckel-colors-typeface-secondary)',
    accent: 'var(--maximeheckel-colors-brand)',
    errorSurface: 'var(--maximeheckel-colors-danger-emphasis)',
    error: 'var(--maximeheckel-colors-danger)',
    surface3: 'var(--maximeheckel-colors-emphasis)',
    surface2: 'var(--maximeheckel-border-color)',
    surface1: 'var(--code-snippet-background)',
  },
  syntax: {
    plain: 'var(--token-comment)',
    comment: {
      color: 'var(--token-comment)',
    },
    keyword: 'var(--token-keyword)',
    tag: 'var(--token-symbol)',
    punctuation: 'var(--token-punctuation)',
    definition: 'var(--token-function)',
    property: 'var(--token-function)',
    static: 'var(--token-comment)',
    string: 'var(--token-selector)',
  },
  font: {
    body: 'var(--font-display)',
    mono: 'var(--font-mono)',
    size: '14px',
    lineHeight: '26px',
  },
};

const defaultEditorOptions = {
  showNavigator: false,
  showInlineErrors: true,
  showLineNumbers: true,
  editorHeight: 520,
};

// Styles
const SandpackWrapper = styled('div', {
  '.sp-layout': {
    position: 'relative',
    marginBottom: '2.25rem',
    borderRadius: 'var(--border-radius-2)',
    boxShadow: Shadows[1],
    '@media (max-width: 750px)': {
      display: 'block',
    },
    '@media (max-width: 1200px)': {
      /**
       * Make it fullbleed!
       */
      width: '100vw',
      left: '50%',
      right: '50%',
      marginLeft: '-50vw',
      marginRight: '-50vw',
      borderRadius: '0px',
    },
    '@media (min-width: 1200px)': {
      position: 'relative',
      width: 'calc(100% + 200px)',
      marginLeft: '-100px',
      marginRight: '-100px',
    },
  },
  '.cm-gutterElement': {
    fontSize: '12px',
    userSelect: 'none',
    opacity: '1',
    color: 'var(--maximeheckel-colors-typeface-tertiary)',
  },
  '.button': {
    backgroundColor: 'var(--maximeheckel-colors-body)!important',
    cursor: 'pointer !important',
    '&:hover': {
      backgroundColor: 'var(--maximeheckel-colors-body)!important',
    },
  },
});

// TODO extends from sandpack type
interface SandpackOptions {
  editorWidthPercentage: number;
  editorHeight: number;
}

interface SandpackProps {
  template: SandpackPredefinedTemplate;
  options?: SandpackOptions;
  // Type using Sandpack built in types
  files: Record<string, any>;
  dependencies?: Record<string, string>;
  autorun?: boolean;
  editorOnly?: boolean;
  renderOnly?: boolean; // TODO Make sure you can't have autorun = true and renderOnly = true
}

const Sandpack = (props: SandpackProps) => {
  const {
    files,
    dependencies,
    template,
    autorun = true,
    editorOnly = false,
    renderOnly = false,
  } = props;
  const editorPart = props.options?.editorWidthPercentage || 50;
  const previewPart = 100 - editorPart;

  const defaultFilesByTemplate: Record<SandpackPredefinedTemplate, any> = {
    react: setupFiles,
    // TODO
    'react-ts': '',
    vanilla: '',
    'vanilla-ts': '',
    angular: '',
    vue: '',
    vue3: '',
    svelte: '',
    solid: '',
  };

  return (
    <SandpackWrapper>
      <SandpackProvider
        template={template}
        theme={theme}
        files={{
          ...files,
          ...defaultFilesByTemplate[template],
        }}
        customSetup={{
          dependencies: dependencies || {},
        }}
        options={{
          autorun,
        }}
      >
        <SandpackLayout>
          {!editorOnly ? (
            <SandpackPreview
              showNavigator={defaultEditorOptions.showNavigator}
              style={{
                height: defaultEditorOptions.editorHeight,
                flexGrow: previewPart,
                flexShrink: previewPart,
                minWidth: 700 * (previewPart / (previewPart + editorPart)),
              }}
            />
          ) : null}
          {!renderOnly ? (
            <SandpackCodeEditor
              {...defaultEditorOptions}
              style={{
                height: defaultEditorOptions.editorHeight,
                flexGrow: editorPart,
                flexShrink: editorPart,
                minWidth: 700 * (editorPart / (previewPart + editorPart)),
              }}
            />
          ) : null}
        </SandpackLayout>
      </SandpackProvider>
    </SandpackWrapper>
  );
};

export default Sandpack;
