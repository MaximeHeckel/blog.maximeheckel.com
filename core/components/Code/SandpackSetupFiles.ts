const IndexJs = `import React, { StrictMode } from "react";
import { createRoot } from 'react-dom/client';
import "./styles.css";
import "./center.css";

import App from "./App";

const container = document.getElementById('root');
const root = createRoot(container);
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);
`;

const StylesCss = `/* Custom styles for your React previews here */
body {
  font-family: sans-serif;
  -webkit-font-smoothing: auto;
  -moz-font-smoothing: auto;
  -moz-osx-font-smoothing: grayscale;
  font-smoothing: auto;
  text-rendering: optimizeLegibility;
  font-smooth: always;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
}
`;

const CenterFlexCss = `
body {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 450px;
    color: currentColor;
}
`;

const IndexHtml = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>React app</title>
</head>
<body>
  <div id="root"></div>
</body>
</html>
`;

const setupFiles = {
  '/index.js': {
    code: IndexJs,
    hidden: true,
  },
  '/center.css': {
    code: CenterFlexCss,
    hidden: true,
  },
  '/styles.css': {
    code: StylesCss,
    hidden: true,
  },
  '/index.html': {
    code: IndexHtml,
    hidden: true,
  },
};

export default setupFiles;
