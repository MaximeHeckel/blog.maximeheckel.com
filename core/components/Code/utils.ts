import { Language } from 'prism-react-renderer';
import { JSXElementConstructor, ReactElement } from 'react';
import { PrePropsType } from './types';

export const preToCodeBlock = (preProps: PrePropsType) => {
  const children = preProps.children as
    | ReactElement<any, string | JSXElementConstructor<any>>
    | undefined;

  if (children && children.props) {
    const { children: codeString, className = '', ...props } = children.props;

    const matches = className.match(/language-(?<lang>.*)/);
    return {
      className,
      codeString: codeString.trim(),
      language:
        matches && matches.groups && matches.groups.lang
          ? (matches.groups.lang as Language)
          : ('' as Language),
      ...props,
    };
  }
};

const RE = /{([\d,-]+)}/;

export const calculateLinesToHighlight = (metastring: string | null) => {
  if (!metastring || !RE.test(metastring)) {
    return () => false;
  } else {
    const lineNumbers = RE.exec(metastring)![1]
      .split(',')
      .map((v) => v.split('-').map((val) => parseInt(val, 10)));
    return (index: number) => {
      const lineNumber = index + 1;
      const inRange = lineNumbers.some(([start, end]) =>
        end ? lineNumber >= start && lineNumber <= end : lineNumber === start
      );
      return inRange;
    };
  }
};

const RETitle = /title=[A-Za-z](.+)/;

export const hasTitle = (metastring: string | null) => {
  if (!metastring || !RETitle.test(metastring)) {
    return '';
  } else {
    return RETitle.exec(metastring)![0].split('title=')[1];
  }
};
