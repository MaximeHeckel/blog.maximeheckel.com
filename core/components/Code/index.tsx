import dynamic from 'next/dynamic';
import React from 'react';
import { PrePropsType } from './types';
import { preToCodeBlock } from './utils';

const CodeBlock = dynamic(() => import('./CodeBlock'));

const Code = (preProps: PrePropsType) => {
  const props = preToCodeBlock(preProps);

  if (props) {
    return <CodeBlock {...props} />;
  }

  return null;
};

export default Code;
