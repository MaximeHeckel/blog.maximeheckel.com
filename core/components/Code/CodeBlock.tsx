import { Card, styled } from '@maximeheckel/design-system';
import Highlight, { Prism, defaultProps } from 'prism-react-renderer';
import CopyToClipboardButton from '../Buttons/CopyToClipboardButton';
import { CodeBlockProps, HighlightedCodeTextProps } from './types';
import { calculateLinesToHighlight, hasTitle } from './utils';

// @ts-ignore
(typeof global !== 'undefined' ? global : window).Prism = Prism;

/**
 * This imports the syntax highlighting style for the Swift and GLSLlanguage
 */
require('prismjs/components/prism-swift');
require('prismjs/components/prism-glsl');

export const HighlightedCodeText = (props: HighlightedCodeTextProps) => {
  const { codeString, language, highlightLine } = props;

  return (
    <Highlight
      {...defaultProps}
      theme={{ plain: {}, styles: [] }}
      code={codeString}
      // @ts-ignore let glsl be a valid language
      language={language}
    >
      {({ className, style, tokens, getLineProps, getTokenProps }) => (
        <Pre className={className} style={style}>
          {tokens.map((line, index) => {
            const { className: lineClassName } = getLineProps({
              className:
                highlightLine && highlightLine(index) ? 'highlight-line' : '',
              key: index,
              line,
            });

            return (
              <Line
                data-testid={
                  highlightLine && highlightLine(index)
                    ? 'highlight-line'
                    : 'line'
                }
                key={index}
                className={lineClassName}
              >
                <LineNo data-testid="number-line">{index + 1}</LineNo>
                <LineContent>
                  {line.map((token, key) => {
                    return (
                      <span
                        data-testid="content-line"
                        key={key}
                        {...getTokenProps({
                          key,
                          token,
                        })}
                      />
                    );
                  })}
                </LineContent>
              </Line>
            );
          })}
        </Pre>
      )}
    </Highlight>
  );
};

const CodeBlock = (props: CodeBlockProps) => {
  const { codeString, language, metastring } = props;

  const highlightLineFn = calculateLinesToHighlight(metastring);
  const title = hasTitle(metastring);

  return (
    <Card
      css={{
        marginBottom: '32px',
        background: 'unset',

        '@media(max-width: 750px)': {
          /**
           * Make it fullbleed!
           */
          width: '100vw',
          position: 'relative',
          left: '50%',
          right: '50%',
          marginLeft: '-50vw',
          marginRight: '-50vw',
          borderRadius: '0px',
        },
      }}
    >
      {title ? (
        <Card.Header
          css={{
            padding: '0px 16px',
            backgroundColor: 'var(--code-snippet-background)',
          }}
        >
          <CodeSnippetTitle data-testid="codesnippet-title">
            {title}
          </CodeSnippetTitle>
          <CopyToClipboardButton title={title} text={codeString} />
        </Card.Header>
      ) : null}
      <HighlightedCodeText
        codeString={codeString}
        language={language}
        highlightLine={highlightLineFn}
      />
    </Card>
  );
};

export default CodeBlock;

const Pre = styled('pre', {
  marginTop: '0',
  marginBottom: '0',
  textAlign: 'left',
  padding: '8px 0px',
  overflow: 'auto',
  borderBottomLeftRadius: 'var(--border-radius-2)',
  borderBottomRightRadius: 'var(--border-radius-2)',
  backgroundColor: 'var(--code-snippet-background)',
  fontFamily: 'var(--font-mono)',
  fontSize: 'var(--font-size-1)',
  lineHeight: '26px',

  '.token.parameter,.token.imports,.token.plain,.token.comment,.token.prolog,.token.doctype,.token.cdata': {
    color: 'var(--token-comment)',
  },

  '.token.punctuation': {
    color: 'var(--token-punctuation)',
  },

  '.token.property,.token.tag,.token.boolean,.token.number,.token.constant,.token.symbol,.token.deleted': {
    color: 'var(--token-symbol)',
  },

  '.token.selector,.token.attr-name,.token.char,.token.builtin,.token.number,.token.string,.token.inserted': {
    color: 'var(--token-selector)',
  },

  '.token.operator,.token.entity,.token.url,.language-css .style': {
    color: 'var(--token-operator)',
  },

  '.token.atrule,.token.attr-value,.token.keyword': {
    color: 'var(--token-keyword)',
  },

  '.token.function,.token.maybe-class-name,.token.class-name': {
    color: 'var(--token-function)',
  },

  '.token.regex,.token.important,.token.variable': {
    color: 'var(--token-operator)',
  },
});

const Line = styled('div', {
  display: 'table',
  borderCollapse: 'collapse',
  padding: '0px 14px',
  borderLeft: '3px solid transparent',

  '&.highlight-line': {
    background: 'var(--maximeheckel-colors-emphasis)',
    borderColor: 'var(--maximeheckel-colors-brand)',
  },

  '&:hover': {
    backgroundColor: 'var(--maximeheckel-colors-emphasis)',
  },
});

const LineNo = styled('div', {
  width: '45px',
  padding: '0 12px',
  userSelect: 'none',
  opacity: '1',
  color: 'var(--maximeheckel-colors-typeface-tertiary)',
});

const LineContent = styled('span', {
  display: 'table-cell',
  width: '100%',
});

const CodeSnippetTitle = styled('p', {
  marginBlockStart: '0px',
  fontSize: 'var(--font-size-1)',
  marginBottom: '0px',
  color: 'var(--maximeheckel-colors-typeface-primary)',
  fontWeight: '500',
});
