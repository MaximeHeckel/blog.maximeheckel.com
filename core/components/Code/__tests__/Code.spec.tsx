/* eslint-disable react/no-unescaped-entities */
import preloadAll from 'jest-next-dynamic';
import { cleanup, render } from '@testing-library/react';
import React from 'react';
import Code from '../';
import { preToCodeBlock, calculateLinesToHighlight, hasTitle } from '../utils';

beforeAll(async () => {
  await preloadAll();
});

afterEach(() => {
  cleanup();
});

describe('Code', () => {
  it('hasTitle returns the title part of a given metastring if present', () => {
    expect(hasTitle('hello,world,title=Test123')).toBe('Test123');
    expect(hasTitle('hello,world')).toBe('');
  });

  it('calculateLinesToHighlight returns the proper set of lignes to highlight of a given metastring', () => {
    expect(
      calculateLinesToHighlight(
        'javascript {7-12} title=Writing the rule (step 1)'
      )(1)
    ).toBeFalsy();
    expect(
      calculateLinesToHighlight(
        'javascript {7-12} title=Writing the rule (step 1)'
      )(7)
    ).toBeTruthy();

    expect(
      calculateLinesToHighlight('javascript title=Writing the rule (step 1)')(7)
    ).toBeFalsy();
  });

  it('preToCodeBlock returns the proper set of props for our code block components if the children are of type <code>', () => {
    const preProps = {
      children: {
        props: {
          children: 'some code to render',
          mdxType: 'code',
          metastring: 'javascript',
        },
      },
    };
    expect(preToCodeBlock(preProps)).toMatchSnapshot();
  });

  it('Renders a Codeblock component when the proper preProps are passed', () => {
    const { container, getByTestId } = render(
      <Code>
        <div metastring="javascript">var hello="world"</div>
      </Code>
    );

    expect(container.querySelector('pre[class="prism-code"]')).toBeDefined();
    expect(getByTestId('number-line')).toBeDefined();

    expect(getByTestId('number-line')).toHaveTextContent(1);
  });

  it('Renders a Codeblock with title when the proper preProps are passed', () => {
    const { container, getByTestId } = render(
      <Code>
        <div metastring="javascript title=test">some code to render</div>
      </Code>
    );

    expect(
      container.querySelector('p[data-testid="codesnippet-title"]')
    ).toHaveTextContent('test');
    expect(container.querySelector('pre[class="prism-code"]')).toBeDefined();
    expect(getByTestId('number-line')).toBeDefined();
    expect(getByTestId('number-line')).toHaveTextContent(1);
    expect(container.querySelector('button')).toBeInTheDocument();
  });

  it('Renders a Codeblock with title and line highlight when the proper preProps are passed', () => {
    const { container, getAllByTestId } = render(
      <Code>
        <div metastring="javascript {1-3} title=test">
          {`some code to render
            some code to render 2
            some code to render 3
            some code to render 4`}
        </div>
      </Code>
    );

    expect(
      container.querySelector('p[data-testid="codesnippet-title"]')
    ).toHaveTextContent('test');
    expect(getAllByTestId('number-line')).toHaveLength(4);
    expect(getAllByTestId('highlight-line')).toHaveLength(3);
    expect(container.querySelector('button')).toBeInTheDocument();
  });
});
