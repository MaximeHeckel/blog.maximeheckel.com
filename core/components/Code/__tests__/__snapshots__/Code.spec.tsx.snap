// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Code > preToCodeBlock returns the proper set of props for our code block components if the children are of type <code> 1`] = `
{
  "className": "",
  "codeString": "some code to render",
  "language": "",
  "mdxType": "code",
  "metastring": "javascript",
}
`;
